data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,307 @@
+ */
+ListNode *reverseLL(ListNode *head)
+{
+    if (!head)
+        return NULL;
+
+    ListNode *new_head = NULL;
+    ListNode *last = head->prev;
+    ListNode *curr = last, *prev;
+
+    // traverse list in backward direction
+    while (curr->prev != last)
+    {
+        prev = curr->prev;
+        new_head = insert_at_last(new_head, curr->value);
+        curr = prev;
+    }
+    new_head = insert_at_last(new_head, curr->value);
+
+    return new_head;
+}
+
+/**
+ * @brief The function that will return current size of list
+ *
+ * @param head      starting pointer of list
+ * @return int      size of list
+ */
+int getsize(ListNode *head)
+{
+    int size = 1;
+    ListNode *temp = head->next;
+    while (temp != head)
+    {
+        temp = temp->next;
+        size++;
+    }
+    return size;
+}
+
+/**
+ * @brief   Display list function
+ * @param   head   starting pointer of list
+ * @returns void
+ */
+
+void display(ListNode *head)
+{
+    printf("\nIts Your Linked-list\n\n");
+    ListNode *temp;
+    temp = head;
+    if (head != NULL)
+    {
+        while (temp->next != head)
+        {
+            printf("%lld -> ", temp->value);
+            temp = temp->next;
+        }
+        if (temp->next == head)
+        {
+            printf("%lld", temp->value);
+        }
+    }
+    else
+    {
+        printf("list is Empty\n");
+    }
+    printf("\n");
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    // Creating an empty list
+    ListNode *head = NULL;
+    // Flg to stop loop and value to insert
+    int flag = 1;
+    uint64_t value = 0;
+    while (flag == 1)
+    {
+        int ch;
+        printf("\nWhat do you want to perform?\n");
+        printf(
+            "ENTER YOUR CHOICE\n1. Insert At head\n2. Insert At End\n3. "
+            "Delete First node \n4. Delete Last Node \n5. Reverse List\n6. Get "
+            "List size\n7. Display List"
+            "\n8. Exit\n");
+        scanf("%d", &ch);
+        switch (ch)
+        {
+        case 1:
+            printf("Enter the data: ");
+            scanf("%lld", &value);
+            head = insert_at_head(head, value);
+            break;
+        case 2:
+            printf("Enter the data: ");
+            scanf("%lld", &value);
+            head = insert_at_last(head, value);
+            break;
+        case 3:
+            head = delete_first_node(head);
+            break;
+
+        case 4:
+            head = delete_last_node(head);
+            break;
+
+        case 5:
+            head = reverseLL(head);
+            printf("\nList succesfully reversed\n\n");
+            break;
+
+        case 6:
+            printf("\nSize of list is %d\n\n", getsize(head));
+            break;
+
+        case 7:
+            display(head);
+            break;
+
+        case 8:
+            flag = 0;
+            break;
+
+        default:
+            printf("\nIncorrect choice. Re-Enter values.\n\n");
+        }
+    }
+    return 0;
+}
```

Comment:
Trivia: The C standard actually requires that every non empty source file ends with a newline. Read more about it [here](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file)